semanticdb/metacp/src/main/scala/scala/meta/internal/javacp/Javacp.scala

```diff
@@ -166,12 +166,23 @@ object Javacp {
             method.signature.params
           }
 
-        val parameters: List[s.SymbolInformation] = params.zipWithIndex.map {
-          case (param: JavaTypeSignature, i) =>
-            val paramDisplayName = {
-              if (method.node.parameters == null) "param" + i
-              else method.node.parameters.get(i).name
-            }
+        /* Drop the constructor argument that holds the reference to the outer class.
+         * Previous method doesn't work for JDK 21.
+         */
+        val paramsFiltered = params.zipWithIndex.flatMap { case (param, i) =>
+          val paramDisplayName =
+            Option(method.node.parameters)
+              .flatMap(x => Option(x.get(i)))
```

Comment:
This will do a off-by-one lookup if the head was removed in the block above. It looks like [we build something with JEP-118's `javac -parameters`](https://github.com/scalameta/scalameta/blame/d52c192c229bcc411da61ecc801545cc2feca997/build.sbt#L340) so we could assert on a custom inner class to confirm that.

Here is my untested suggestion to consolidate the "skip synthetic param" & "support absent or null-named params" steps :

```scala
        // The first argument might be a synthetic param holding the reference to the outer class
        val dropFirstArgument = isConstructor && hasOuterClassReference

        val params = method.signature.params.zipWithIndex.collect {
          case (param, i) if !(dropFirstArgument && i == 0) =>
            val paramDisplayNameOpt =
              Option(method.node.parameters)
                .flatMap(x => Option(x.get(i)))
                .flatMap(x => Option(x.name))
            (param, paramDisplayNameOpt)
        }

        val parameters: List[s.SymbolInformation] = params.zipWithIndex.map {
          case ((param: JavaTypeSignature, paramDisplayNameOpt), i) =>
            val paramDisplayName = paramDisplayNameOpt.getOrElse("param" + i)
```

In the first zipWithIndex, i refers to the index in the class file which includes synthetic params, while in the second it does not, keeping `param0` as name of the first non-synthetic param in inner classes constructors.